[offload] Standalone build fixes

mgorny opened the pull request:
A fair number of fixes to get standalone builds of offload working — mostly copying missing bits from openmp. It's almost ready — I still need to figure out why some of the tsts aren't linking to the right libraries.

mgorny committed:
[offload] Include CheckCXXCompilerFlag in CMake

Include CheckCXXCompilerFlag before it is used in the top-level CMake
file.  This fixes standalone builds, but it is also cleaner than
assuming that some previous CMake file will include it for us.

mgorny committed:
[offload] [test] Add "omp" test dependency only when present

Add the `omp` target dependency to tests only when the respective target
is present, i.e. when not building standalone against system libomp.

mgorny committed:
[offload] Define OPENMP_TEST*_FLAGS in standalone builds

Copy the `OPENMP_TEST_FLAGS` and `OPENMP_TEST_OPENMP_FLAGS` from openmp
for standalone builds, as required by the lit configs.

mgorny committed:
[offload] Define LIBOMPTARGET_OPENMP_*_FOLDER in standalone builds

Define `LIBOMPTARGET_OPENMP_HEADER_FOLDER`
and `LIBOMPTARGET_OPENMP_HOST_RTL_FOLDER` needed for testing
in standalone builds.

mgorny committed:
[offload] Fix path to llvm-offload-device-info in standalone build

Fix path to `llvm-offload-device-info` tool in standalone build.
This tool is built as part of offload, so rather than looking it up
as a preinstalled tool, set correct `LLVM_RUNTIME_OUTPUT_INTDIR`
for the `llvm_add_tool()` call, and unconditionally set
`OFFLOAD_DEVICE_INFO_EXECUTABLE` using that directory.

jhuber6 commented on offload/plugins-nextgen/amdgpu/CMakeLists.txt:
```diff
@@ -8,14 +8,19 @@ target_sources(omptarget.rtl.amdgpu PRIVATE src/rtl.cpp)
   message(STATUS "Building AMDGPU plugin for dlopened libhsa")
   target_include_directories(omptarget.rtl.amdgpu PRIVATE dynamic_hsa)
   target_sources(omptarget.rtl.amdgpu PRIVATE dynamic_hsa/hsa.cpp)
-  target_link_libraries(omptarget.rtl.amdgpu PRIVATE LLVMFrontendOffloading)
```
I feel like the LLVM library here should be moved to the common plugin.

jhuber6 commented on offload/cmake/DetectTestCompiler/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+cmake_minimum_required(VERSION 3.20.0)
```
Shouldn't this be a CMake module instead of a separate project?

mgorny replied:
I simply don't know. It's copied 1:1 from openmp.

jhuber6 commented on offload/plugins-nextgen/common/CMakeLists.txt:
```diff
@@ -1,13 +1,16 @@
 # NOTE: Don't try to build `PluginInterface` using `add_llvm_library` because we
 # don't want to export `PluginInterface` while `add_llvm_library` requires that.
-add_library(PluginCommon OBJECT
+add_library(PluginCommon STATIC
```
Why is this static now?

mgorny replied:
As I've noted in the commit message, `llvm_config()` doesn't support object libraries.

jhuber6 replied:
I'm pretty sure that making this static will break something, but I can't remember what it is exactly. The LLVM configs here should probably be done when we make the plugin static library, `add_target_library`.

mgorny replied:
Well, all tests passed for me with this change — but I can alternatively change `llvm_config` to use `PRIVATE` linking for object libraries, but I don't know if that isn't going to break something else.

jhuber6 replied:
I don't think the LLVM profile data thing should go here at least. Maybe @EthanLuisMcDonough knows if it was necessary to put it there. All of the LLVM libs should be handled in `add_target_library`, however we do need the dependency on intrinsics because we depend on it.

This is intentionally supposed to create a single object library that's then used to create the static libraries, we could maybe do it with static but I don't see the point when this isn't where any LLVM logic should go anyway.

jhuber6 reviewed (commented):
Do you know if https://github.com/llvm/llvm-project/pull/117933 will conflict  with anything here? It includes a header file cross-project.

mgorny committed:
[offload] [test] Add missing dependency on llvm-offload-device-info

mgorny commented:
Ok, I think it's ready. The remaining test failures I see are in `amdgcn-amd-amdhsa` tests, and they fail with regular build for me as well. I'm guessing it's because I don't have HSA installed.

jhuber6 commented on offload/plugins-nextgen/common/CMakeLists.txt:
```diff
@@ -7,7 +7,10 @@ add_library(PluginCommon OBJECT
 )
-add_dependencies(PluginCommon intrinsics_gen LLVMProfileData)
+add_dependencies(PluginCommon intrinsics_gen)
+if (TARGET LLVMProfileData)
+  add_dependencies(PluginCommon LLVMProfileData)
+endif()
```
Can we remove this and move it with the others?

mgorny replied:
I'm sorry but I don't understand the "move it with the others" part. I don't really understand why an explicit dependency was added in the first place — I thought linking against the library implies that.

jhuber6 replied:
I don't think it needs to be here, it's a library target so it should go with the other LLVM libraries. Moving it should avoid the need for handling object libraries like this.

mgorny replied:
You mean into `supported_jit_targets`?

jhuber6 replied:
`add_target_library` or whatever it's called. The JIT library linking should probably also be moved there. Looking at it now, I'm not sure how it works since it's doing a private `link_libraries` on an object library, which has no link step.

mgorny committed:
[offload] Move additional plugin dependencies to add_target_library

Move the dependencies on `LLVMFrontendOffloading` and `LLVMProfileData`
from amdgpu and common, respectively, to the base `add_target_library()`
macro.  This means we can take advantage of LINK_COMPONENTS and do not
have to add extra complexity to handle dylib and standalone builds.

mgorny committed:
[cmake] Use DetectTestCompiler from openmp directory

Fix the DetectTestCompiler project use to reference the openmp source
tree, since the respective files were not copied to offload, and there
is no point in duplicating them.

Fixes #90333

mgorny commented:
Moved the dependencies into `add_target_library`, as discussed. Also switched to reusing `DetectTestCompiler` from openmp instead of copying the files here.

arsenm commented on offload/cmake/OpenMPTesting.cmake:
```diff
@@ -136,7 +125,7 @@ if (${OPENMP_STANDALONE_BUILD})
   file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/DetectTestCompiler)
   execute_process(
-    COMMAND ${CMAKE_COMMAND} -G${CMAKE_GENERATOR} ${CMAKE_CURRENT_LIST_DIR}/DetectTestCompiler
+    COMMAND ${CMAKE_COMMAND} -G${CMAKE_GENERATOR} ${CMAKE_SOURCE_DIR}/../openmp/cmake/DetectTestCompiler
```
It's nearly always wrong to use CMAKE_SOURCE_DIR instead of CMAKE_CURRENT_SOURCE_DIR

mgorny replied:
Fixed.

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -286,6 +290,26 @@ if(OPENMP_STANDALONE_BUILD)
   )
+
+  find_path (
+    LIBOMP_INCLUDE_DIR
```
I will forever maintain that only the plugins should've been moved. Maybe it's not too late to do that instead.

mgorny commented:
> What is the understanding of a standalone build here? Would you mind sharing an example CMake invocation? Thank you.

```
cmake -G Ninja .../llvm-project/offload
```

The one indicated by `OPENMP_STANDALONE_BUILD`.

thesamesam reviewed (commented):
No concerns from me, at least.

thesamesam commented:
@jhuber6 Is it OK if we go ahead with this as-is? We can file a bug for reconsidering some of the split if needed, but right now, offload is broken in our configuration and has been for quite some time - and we had some users start requesting it.

jplehr commented:
Can you see if your patch still works or needs some modification after https://github.com/llvm/llvm-project/pull/117933 has landed? Once confirmed, I think we can accept.
Thank you, much appreciated!

I ran this through some buildbot configs and did not see any issues.

jplehr commented:
I personally would appreciate to rebase this PR and add the changes required to work with the libc patch. Don't know how difficult that will be though.

jplehr commented:
> > I personally would appreciate to rebase this PR and add the changes required to work with the libc patch. Don't know how difficult that will be though.
> 
> Actually, the right approach would be to fix the libc path not to use `LLVM_MAIN_SRC_DIR` (after all, it's not using `llvm` subdirectory at all), and instead declare a path relative to `CMAKE_CURRENT_SOURCE_DIR`, much like `LLVM_COMMON_CMAKE_UTILS` is declared now.

We could key off of `devicertl_base_directory` which is already used in those places.

mgorny commented:
Thanks!

Is it okay to merge this now? I'd also like to try backporting it for 19.x (at least for Gentoo, ideally upstream too).

jplehr reviewed (approved):
LGTM
I ran it through a couple of configs locally.
